src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs

```diff
@@ -111,9 +144,194 @@ internal class DiagnosticComputer
                 }
             }
 
-            var diagnosticsComputer = new DiagnosticComputer(document, project,
-                solutionChecksum, ideOptions, span, analysisKind, analyzerInfoCache, hostWorkspaceServices);
-            return diagnosticsComputer.GetDiagnosticsAsync(analyzerIds, reportSuppressedDiagnostics, logPerformanceInfo, getTelemetryInfo, cancellationToken);
+            // We execute explicit, user-invoked diagnostics requests with higher priority compared to implicit requests
+            // from clients such as editor diagnostic tagger to show squiggles, background analysis to populate the error list, etc.
+            var diagnosticsComputer = new DiagnosticComputer(document, project, solutionChecksum, ideOptions, span, analysisKind, analyzerInfoCache, hostWorkspaceServices);
+            return isExplicit
+                ? diagnosticsComputer.GetHighPriorityDiagnosticsAsync(analyzerIds, reportSuppressedDiagnostics, logPerformanceInfo, getTelemetryInfo, cancellationToken)
+                : diagnosticsComputer.GetNormalPriorityDiagnosticsAsync(analyzerIds, reportSuppressedDiagnostics, logPerformanceInfo, getTelemetryInfo, cancellationToken);
+        }
+
+        private async Task<SerializableDiagnosticAnalysisResults> GetHighPriorityDiagnosticsAsync(
+            IEnumerable<string> analyzerIds,
+            bool reportSuppressedDiagnostics,
+            bool logPerformanceInfo,
+            bool getTelemetryInfo,
+            CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Step 1:
+            //  - Create the core 'computeTask' for computing diagnostics.
+            var computeTask = GetDiagnosticsAsync(analyzerIds, reportSuppressedDiagnostics, logPerformanceInfo, getTelemetryInfo, cancellationToken);
+
+            // Step 2:
+            //  - Add this computeTask to the list of currently executing high priority tasks.
+            //    This list of high priority tasks is used in 'GetNormalPriorityDiagnosticsAsync'
+            //    method to ensure that any new or cancelled normal priority task waits for all
+            //    the executing high priority tasks before starting its execution.
+            //  - Note that it is critical to do this step prior to Step 3 below to ensure that
+            //    any canceled normal priority tasks in Step 3 do not resume execution prior to
+            //    completion of this high priority computeTask. 
+            lock (s_gate)
+            {
+                Debug.Assert(!s_highPriorityComputeTasks.Contains(computeTask));
+                s_highPriorityComputeTasks.Add(computeTask);
```

Comment:
if you have a set, you can then assert that Add returned the right value without a linear walk.